Adding RequireJS

clrux opened the pull request:
# Adding RequireJS to the Pattern Library

This large piece of work brings in [RequireJS](http://requirejs.org) and modifies existing JavaScript files to be loaded through it. Additionally, it does some significant refactoring and clean up including:
* removing the concatenation and renaming process from the `gulp` workflow
* modularizing the `ui.js` script which was previously loose functions
* breaking out the `Tabs` function into its own file

---

## Using RequireJS
RequireJS works by `require`(ing) immediately needed files, and then in each subsequently loaded file, `define`(ing) dependencies for each respective file.

This allows us to specify on a per-file basis, the requirements and dependencies for given JS files. For example, in `assets-secondary.html` rather than including the pattern library JS and the documentation site JS, we simply point to RequireJS and give it a `data` file to load (the required files).

In `required.js`, we define those files that need to load up-front (required) and then any dependencies.

In subsequent files that are loaded, __those__ files in turn have their own sets of required files and dependencies. Files are loaded on an as-needed basis, rather than all at once, ultimately saving download overhead, speeding up the site in general, and giving us control over which files are used where.

### Concatenation and minification

Since we need to load each file individually, we can no longer concatenate all of the JS files into single, minified versions. We can still minify them individually though. Instead of bundling all of the files together and minifying them, we're only loading the JS files as they're needed, using their minified versions.

### Immediate load
In the `assets-secondary.html` included partial, RequireJS is being loaded as the _only_ JS file as so:
```
<script data-main="{{ site.baseurl }}public/pldoc/js/required.js" src="{{ site.baseurl }}public/js/require.js"></script>
```

I've also included here a `data-main` attribute which points to the file handler that we want to load when the site loads. So in this case `required.js` defines all of the subsequent files that we want to call.

### Dependencies
To call a script, we use the following:
```
define(["path/to/file"], function(callsign) {});
```

Note that you should omit the `.js` extension as RequireJS assumes it. The `callsign` is a handle (or nickname?) for the file you're loading. You can then reference that file via that handle.

---

__Reviewers:__
- [x] @talbs for overall cleanliness and functionality; make sure things work as expected
- [x] @AlasdairSwan for RequireJS work and assistance

talbs commented:
@clrux, this looks fine from both general PL code structure and local rendering perspectives. Looking forward to @AlasdairSwan (or someone else familiar with RequireJS if he's still swamped on Q4/Q1 stuff).

One last ask, can you update the README and any general general library mentions within our styleguide wiki to include the fact that we're using RequireJS.

:+1: once you add any needed documentation as well as the ENG review is complete.

AlasdairSwan commented on _src/pldoc/static/js/required.js:
```diff
@@ -0,0 +1,20 @@
+    '../../js/svg4everybody.min'
+    ],
+    function($, Ui) {}
+);
```
Can we rename this file something more descriptive than "required"?

AlasdairSwan commented on _src/pldoc/static/js/color-contrast.js:
```diff
@@ -130,5 +132,4 @@ $(function() {
 
     AccessibilityColorContrast.init();
-
 });
```
I don't remember reviewing this file. There are some examples of 
```
var that = this;
```
which should be cleaned up and some variables not getting declared at the tops of functions.

AlasdairSwan replied:
L87 don't need '=== true'

clrux replied:
True, because if it's set, it is true. Thanks.

AlasdairSwan commented on _src/pldoc/static/js/size-slider.js:
```diff
@@ -55,5 +57,4 @@ $(function() {
 
     IconFontSliderControl.init();
-
 });
```
I didn't review this one yet either.

clrux replied:
I actually do need the object brackets. It fails to run without them.

AlasdairSwan replied:
weird, that's fine 'tho.

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,49 @@
+
+        resetInterface: function(wrapper) {
+            wrapper.find('.pldoc-link').each(function() {
+                $(this).removeClass(Tabs.vars.activeClass);
```
please replace $(this) with passed in value from function

clrux replied:
`$(this)` is the jQuery object in each iteration of the loop. The `event` being passed is the array key (i.e., 0, 1, 2, 3, etc.). What did you have in mind here?

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,49 @@
+            });
+
+            wrapper.find(Tabs.vars.panel).each(function() {
+                $(this).removeClass(Tabs.vars.activeClass).addClass(Tabs.vars.hiddenClass);
```
please replace $(this) with passed in value from function
can you also please take a new line for .addClass

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,49 @@
+                e.preventDefault();
+
+                var content = $(this).attr('href');
+
```
Please replace $(this) with $(e.currentTarget)

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,49 @@
+            Tabs.vars.tab.find('.pldoc-link').on('click', function(e) {
+                e.preventDefault();
+
+                var content = $(this).attr('href');
```
All variables need to be declared at the top of the function (because hoisting)

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,49 @@
+
+        makeActive: function(tab, content) {
+            tab.addClass(Tabs.vars.activeClass);
+            $(content).addClass(Tabs.vars.activeClass).removeClass(Tabs.vars.hiddenClass);
```
new line for .removeClass please

AlasdairSwan commented on _src/pldoc/static/js/ui.js:
```diff
@@ -1,142 +1,111 @@
-    });
-});
+    Ui.init();
+});
```
Please replace all instances of this where possible

clrux replied:
Replace with something? Or remove?

AlasdairSwan replied:
Use alternatives to
```
var that = this;
```

clrux replied:
Oh, I thought you were talking about the specific line and was confused. Okay, on it.

AlasdairSwan replied:
Sorry, I see the confusion. Apologies, I will use more words next time :)

AlasdairSwan commented on _src/pldoc/static/js/ui.js:
```diff
@@ -1,142 +1,111 @@
-            range.selectNodeContents(element);
-            selection.removeAllRanges();
-            selection.addRange(range);
+        selectText: function(that) {
```
Can you please rename the passed in variable to something more meaningful

AlasdairSwan commented on _src/pldoc/static/js/ui.js:
```diff
@@ -1,142 +1,111 @@
+                range.select();
+            } else if (window.getSelection) {
+                var selection = window.getSelection();
+                var range = doc.createRange();
```
vars should be declared at the top of the function

AlasdairSwan commented on _src/pldoc/static/js/ui.js:
```diff
@@ -1,142 +1,111 @@
+                element = that[0];
+
+            if (doc.body.createTextRange) {
+                var range = doc.body.createTextRange();
```
vars should be declared at the top of the function

AlasdairSwan commented:
@clrux please clean up the gulp files so at the top there is only one var declaration

clrux commented:
@AlasdairSwan I've gone through and addressed most of your feedback. I did have a few questions though, if you wouldn't mind taking another spin?

AlasdairSwan commented on _src/pldoc/static/js/color-contrast.js:
```diff
@@ -36,20 +37,19 @@ $(function() {
+                bg = AccessibilityColorContrast.getL( AccessibilityColorContrast.rgbaToHex( $(this).css('backgroundColor') ) );
+                fg = AccessibilityColorContrast.getL( AccessibilityColorContrast.rgbaToHex( $(this).find('.color-class').css('color') ) );
 
                 var ratio = (Math.max(bg, fg) + 0.05) / (Math.min(bg, fg) + 0.05),
```
All variables should be declared at the top of the function

AlasdairSwan commented on _src/pldoc/static/js/size-slider.js:
```diff
@@ -22,38 +24,30 @@ $(function() {
+                // console.log(event.target.value);
+                IconFontSliderControl.updateIconSize(event.target.value);
+                IconFontSliderControl.updateAria(event.target.value);
+                IconFontSliderControl.updateInputText(event.target.value);
```
You use event.target.value 3 times so it would make sense to set it as a var at the top of the function and reference that instead.

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,54 @@
+
+        resetInterface: function(wrapper) {
+            wrapper.find('.pldoc-link').each(function(event) {
+                $(this).removeClass(Tabs.vars.activeClass);
```
replace this with event.currentTarget

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,54 @@
+            });
+
+            wrapper.find(Tabs.vars.panel).each(function() {
+                $(this).removeClass(Tabs.vars.activeClass)
```
replace this with value passed into each function

AlasdairSwan commented on _src/pldoc/static/js/tabs.js:
```diff
@@ -0,0 +1,54 @@
+        },
+
+        handleTabClick: function() {
+            var el, content;
```
these vars are only used within the .on function so their declaration should be moved there.

AlasdairSwan commented:
@clrux I reviewed and there are 3 files that need minor changes. Also, a note on consistency. For passed in events you sometimes use 'e' and sometimes 'event', for ease of reading could you please always use the full 'event' variables. When assigning jQuery object to variables it would be good if you could always prefix the var with $.

Once these changes are made we are good to go :+1:

clrux commented:
Thanks @AlasdairSwan. I've moved the variables, replaced `this` with `event.currentTarget` and went through and changed all `e` to `event` to be consistent.

clrux commented:
Merging with thumbs from both @AlasdairSwan and @talbs.

talbs commented:
FYI, @andy-armstrong
